### ISSUE TITLE
Example parsers

### ISSUE THREAD

**Issue by Geal** [Geal/nom#14]
We currently have a few example parsers. In order to test the project and make it useful, other formats should be implemented. Here is a list, if anyone wants to try it:

- **text file formats**:
  - [x] INI
  - [ ] YAML
  - [ ] TOML
- **video and image formats**:
  - [x] MP4 (partial implementation)
  - [ ] OGG
  - [ ] MPEG TS
  - [ ] AVI
  - [ ] PNG
  - [ ] JPEG
  - [ ] GIF
- **executable formats**:
  - [ ] Portable executables (PE)
  - [ ] ELF
- **document formats**:
  - [ ] PDF
  - [ ] MS-CFB (compound format, used in doc, xls, ppt, cab, msi files)
  - [ ] TAR
  - [ ] GZ
  - [ ] ZIP
  - [ ] RAR
  - [ ] binary PLIST
- **network packet formats**:
  - [ ] IP
  - [ ] TCP
  - [ ] UDP
  - [ ] NTP
  - [ ] DNS
  - [ ] PCAP
- **crypto related**:
  - [ ] X.509 certificates
  - [ ] DER public and private keys
  - [ ] SSL/TLS packets
  - [ ] OpenPGP

**Comment by thehydroimpulse** [comment]
I'm writing a Thrift library for Rust that'll use Nom for both their IDL and the network protocol, so that can be another example (although in a different repo).

**Comment by Geal** [comment]
Nice idea, that will be useful! Please notify me when it is done, I will add a link in this list.

**Comment by filipegoncalves** [comment]
This looks interesting. Is anyone actively working on any of these parsers? I'd like to work on a few of these.

**Comment by Geal** [comment]
I have some code for a GIF one at https://github.com/Geal/gif.rs but it is hard to test, since the graphical tools in Piston change a lot.

You can pick any of them. Network packets may be the easiest, since they don't require a decompression phase.

I am using the gif example to see what kind of API can be built over nom. Most of the parsing example are done as one pass over the data, but often there is some logic on the side, and it is not easy to encode correctly.

**Comment by elij** [comment]
I've started a fastq (http://en.wikipedia.org/wiki/FASTQ_format) parser https://github.com/elij/fastq.rs

**Comment by Geal** [comment]
@elij this is a great idea! Was it easy to do?

**Comment by elij** [comment]
yup it's a great framework -- though I struggled a bit with eof so I borrowed some code from rust-config (https://github.com/elij/fastq.rs/blob/master/src/parser.rs#L69) -- is there a better solution?

**Comment by Geal** [comment]
yes, eof should be a parser provided by nom, I am just waiting for @filipegoncalves to send a PR :wink:

**Comment by filipegoncalves** [comment]
Hah, sorry for my silence. I've been busy lately. I just sent a PR.

I will be working on one of these example parsers as soon as I get some spare time. There are some great ideas in there!

**Comment by Keruspe** [comment]
I might give tar a try

**Comment by nelsonjchen** [comment]
Does this check off PCAP?

https://github.com/richo/pcapng-rs

**Comment by Geal** [comment]
pcap-ng and pcap are two different formats, right? It seems the consensus now is to move everything to pcap-ng, though.

**Comment by TechnoMancer** [comment]
I will try a FLAC parser, need to add quite a few things for it though.

**Comment by badboy** [comment]
ISO8601 is done in https://github.com/badboy/iso8601 (I hope it's mostly correct.)

**Comment by Geal** [comment]
ok, it should be up to date. More to come :smile:

**Comment by sbeckeriv** [comment]
WARC file format released. https://crates.io/crates/warc_parser

**Comment by Geal** [comment]
@sbeckeriv great, thanks!

**Comment by porglezomp** [comment]
It might be informative to try parsing the rust grammar with nom, if nobody has yet. In any case, I'd like to see a few programming languages on that list, since that's my use case.

**Comment by Geal** [comment]
@porglezomp programming languages examples would definitely be useful, but the Rust grammar might be a bit too much for the first attempt. Which other languages would you like to handle?

**Comment by porglezomp** [comment]
Yeah, I'm aware of the scale problem of Rust. I don't want to write that one, but I think it's a good holy grail for any parser library written in Rust. I'd like to try parsing the Lua grammar first, I think.

**Comment by Geal** [comment]
ok, I added them to the list :)

**Comment by chriskrycho** [comment]
You have INI marked as done; do you have a link to it? (I'd *love* to use this for some tooling I'm hoping to build in 2016; need a good non-trivial example for it, though.)

**Comment by badboy** [comment]
@chriskrycho: https://github.com/Geal/nom/blob/master/tests/ini.rs

**Comment by chriskrycho** [comment]
Thanks very much, @badboy!

**Comment by fbernier** [comment]
I'll give the TOML parser a try very soon.

**Comment by Geal** [comment]
Actually, I think I should rewrite that INI parser, now that more convenient combinators are available.
Also, I should really work on that combinator for space separated stuff

**Comment by Geal** [comment]
@fbernier great! Please keep me posted!

**Comment by l0calh05t** [comment]
Maybe add a simple example for trailing commas in lists? Python has those, but is quite complex. Can't think of a simple example though.

**Comment by johshoff** [comment]
That IRC example is no longer using nom. The parser was moved into its own repository: https://github.com/Detegr/RBot-parser

**Comment by Geal** [comment]
@l0calh05t to parse something like `[a,b,c,]` or `[a,b,c]` ?
@johshoff fixed, thanks

**Comment by l0calh05t** [comment]
@Geal yes, exactly

**Comment by Geal** [comment]
@l0calh05t for `[a,b,c]`, you can parse with `delimited!(char!('['), separated_list!( char!(','), alphabetic), char!(']'))`.
For `[a,b,c,]`, you can have  `delimited!(char!('['), terminated!(alphabetic, char!(',')), char!(']'))`.

A parser that would handle both cases is much trickier.

**Comment by l0calh05t** [comment]
Both is really the more interesting case. And what is needed for Python for example

**Comment by Keruspe** [comment]
Could there be something like maybe_char!(",") which would read a char, consume it if it's ',' or backtrack if it isn't?

**Comment by l0calh05t** [comment]
Problem is that wont work unless a look ahead of more than one character is added automatically

**Comment by Geal** [comment]
In fact, it is easier than I thought, but requires some work:
```rust
preceded!(
  char!('['),
  terminated!(
    separated_list!(
      char!(','),
      alphabetic
    ),
    terminated!(
      opt!(char!(',')),
      char!(']')
    )
  )
)
```

**Comment by badboy** [comment]
I wrote a simplistic [bencode](https://wiki.theory.org/BitTorrentSpecification#Bencoding) parser: [nom-bencode](https://github.com/badboy/nom-bencode).

Not sure if it covers everything (yet).

**Comment by passcod** [comment]
I've started a TOML parser as a learning project: https://github.com/passcod/noml

**Comment by sourrust** [comment]
At this point I'm ready to share my [flac] implementation as an example parser.

[flac]: https://github.com/sourrust/flac

**Comment by joeblew99** [comment]
Hey sourrust. Nom looks great way to do this.
I am interested in parsing different video formats with nom. If there is some existing rust kibs in this space that anyone knows then u could start porting some to nom. Worth a crack to see how it goes.

I am very curious about using the streaming capabilities of nom. For my use case I want to stream data between servers, manipulate frames, and then fan it back into the main stream. 
I would love to get some feedback on some potential gotchas.
Doing this type of work should ultimately feedback into making nom better.

**Comment by tomjakubowski** [comment]
Correct me if I'm wrong, but the linked Redis project doesn't seem to use nom.

**Comment by porglezomp** [comment]
I agree, I checked the history of it's Cargo.toml and at no point was nom listed as a dependency. I'm not sure how it ended up on the list, but it looks like it should be taken off.

**Comment by badboy** [comment]
It does in another branch, which is still not merged because time.

**Comment by thehydroimpulse** [comment]
@Geal you can remove my Thrift library as an example as I'm no longer using Nom in it.

**Comment by joelself** [comment]
I've released a TOML parser. It doesn't let you modify everything possible in the document or create documents from scratch, but does correctly parse TOML, report errors, allow some modification and then output the document with comments and whitespace intact.

**Comment by moosingin3space** [comment]
I've started working on a parser for IP, TCP, UDP, and Ethernet headers.

**Comment by lyze** [comment]
[Java class file parser!](https://github.com/maxmcc/rust-jvm/tree/master/src/parser) It is part of a larger class project. 

The parser uses [helper macros](https://github.com/maxmcc/rust-jvm/blob/master/src/parser/nom_support.rs) based on #160 to get more backtracking support.

**Comment by pwoolcoc** [comment]
Not sure if it's worth putting here or not, but I'm using `nom` to parse strings for the [tracery](https://github.com/galaxykate/tracery) library I am writing for rust: https://github.com/pwoolcoc/tracery-rs

**Comment by jethrogb** [comment]
A subset of C, namely C literals and expressions: https://crates.io/crates/cexpr

**Comment by gz** [comment]
FYI I used nom for a parser of the linux perf data format (https://github.com/gz/rust-perfcnt/blob/master/src/linux/parser.rs) in case you want to add it. In comparison to most examples listed here it parses binary data.

**Comment by nickbabcock** [comment]
[Boxcars](https://github.com/nickbabcock/boxcars) is an example of a [Rocket League](http://www.rocketleaguegame.com/) replay parser with [serde](https://github.com/serde-rs/serde) serialization. Let boxcars be a good example of Rust code using nom, and serde as extensive examples are hard to come by. While lacking user friendly error message -- [among other issues](https://github.com/nickbabcock/boxcars/issues), tests and documentation strive to be thorough.

**Comment by dtolnay** [comment]
> Yeah, I'm aware of the scale problem of Rust. I don't want to write that one, but I think it's a good holy grail for any parser library written in Rust.

As of version 0.10.0, [`syn`](https://github.com/dtolnay/syn) is now able to parse practically all of Rust syntax. One of my test cases is to parse the entire [github.com/rust-lang/rust](https://github.com/rust-lang/rust) repo into an AST and print it back out, asserting that the output is identical to the original.

I am technically not using nom but instead a fork which removes the IResult::Incomplete variant. I found that the extra macro code generated to handle Incomplete was more than doubling the compile time for something that I didn't even want. Nevertheless, the code is enough like nom that I think we can check off the box.

Example snippet to parse one arm of a `match` expression:

```rust
named!(match_arm -> Arm, do_parse!(
    attrs: many0!(outer_attr) >>
    pats: separated_nonempty_list!(punct!("|"), pat) >>
    guard: option!(preceded!(keyword!("if"), expr)) >>
    punct!("=>") >>
    body: alt!(
        map!(block, |blk| ExprKind::Block(BlockCheckMode::Default, blk).into())
        |
        expr
    ) >>
    (Arm {
        attrs: attrs,
        pats: pats,
        guard: guard.map(Box::new),
        body: Box::new(body),
    })
));
```

**Comment by Geal** [comment]
@dtolnay syn is an amazing example, thanks for your hard work :)

**Comment by Geal** [comment]
@dtolnay could I get your input on #356? It might fix your issues with compile times, so I'd like to get your thoughts on this.

**Comment by J-F-Liu** [comment]
I am writing a PDF library using nom to parse PDF syntax. Released v0.1.0 just now.
https://github.com/J-F-Liu/lopdf

**Comment by valarauca** [comment]
So I've implemented a EDI parser for the ANS standard EDI for work with this. Awesome library really useful. Sadly that's owned by my employer.

I've started implementing an x64 assembler with nom. I'm really struggling with writing the parser. The main reason is register names have a lot of overlap, and are very short. For example `r8`, `r8w`,  `r11`, and `r12d`. Ideally I want to map these to an enum. `map!()` makes this easy, but how can I match those terms *in nom*?

**Comment by Keruspe** [comment]
I converted several "keys" to enum values in my brainfuck parser, might or might not be relevant to your needs. See the first parsers defined with "named!" https://github.com/Keruspe/brainfuck.rs/blob/master/src/parser.rs

**Comment by przygienda** [comment]
is there a way (or it would be great if it's possible) to generate EBNF from this? Great package BTW ...

**Comment by ithinuel** [comment]
Hi,
I just pushed a pcap parser : https://github.com/ithinuel/pcap-rs.
It still needs the PR #492 to be merge so it can use official nom crate.

Any feedback is welcome.

**Comment by bbqsrc** [comment]
A parser for the Mediawiki format would be quite useful.

**Comment by dwerner** [comment]
@Geal thanks for an awesome library! I wrote a wavefront obj/mtl 3d mesh parser using it [nom-obj](https://github.com/dwerner/nom-obj), which I published to crates.io

**Comment by olivren** [comment]
I wrote a [parser for the simple key/value text format `.properties`](https://bitbucket.org/olivren/dotproperties), which is a standard for Java configuration files. It uses nom 3.1. Can it be added to the list?

This is the first parser I wrote using a Parser Combinator library. If anyone can review my code I would be delighted. Also, I tried to add error reporting to my code, but I gave up after I tried to insert `add_return_error` and `return_error`  calls all over the place to no avail (in the branch "error-reporting"). Is there an example of a text parser that reports parsing errors?

**Comment by santifa** [comment]
@Geal  thanks for this library.
I've implemented a parser for [URI's](https://github.com/santifa/rrp/blob/master/src/uri.rs) which is 
part of a larger side project for RDF (n3, ttl,...) parsers. The full abnf of rfc 3986 is implemented but the pct-encoding is still a bit messy.

**Comment by dbrgn** [comment]
Here's a parser for ICE candidates SDP (RFC 5245), used for example in WebRTC: https://github.com/dbrgn/candidateparser

**Comment by kamarkiewicz** [comment]
I wrote a Session Initiation Protocol (RFC3261) low-level push parser with API inspired by seanmonstar/httparse (hyper's HTTP parser):
https://github.com/kamarkiewicz/parsip

**Comment by thejpster** [comment]
I'd be interested in something that could parse SNMP MIB and YANG.

https://en.wikipedia.org/wiki/YANG

**Comment by ctrlcctrlv** [comment]
The [BitTorrent example has been deleted](https://github.com/jag426/bittorrent), it seems.

**Comment by Riduidel** [comment]
As a beginner in Rust world, I'm quite sure I will say something horribly wrong, but is there any planned support for some XML dialects ? (typically RSS/ATOM) ?

**Comment by porglezomp** [comment]
@Riduidel if you're specifically interested in just having parsers for those formats, look at https://github.com/rust-syndication. I don't think there's any nom involved there though.

**Comment by mitchtbaum** [comment]
HTTP: https://github.com/hjr3/weldr/blob/00481f80ae60bd6b312805245c126c168ab77b36/src/http/parser.rs

**Comment by ProgVal** [comment]
I wrote a Python parser: https://docs.rs/python-parser/

**Comment by idursun** [comment]
I think [Redis database file format](https://github.com/badboy/rdb-rs) parser is not using nom at all. I couldn't find any reference to nom anywhere.

**Comment by nelsonjchen** [comment]
@idursun Maybe it refers to this old branch from a year before the last update to master. https://github.com/badboy/rdb-rs/tree/nom-parser

**Comment by saggit** [comment]
is there any SQL parser？

**Comment by naturallymitchell** [comment]
> is there any SQL parser？


it'd seem better to me to import it to an sql engine and interact with that data using Diesel.  parsing flat sql files seems very limited.

instead of writing a one-off Rust app to do this, you could add diesel bindings to [Torchbear](https://github.com/foundpatterns/torchbear), see https://github.com/foundpatterns/torchbear/issues/85 , then make a [Speakeasy](https://github.com/foundpatterns/speakeasy) library for transporting data from your schema using content model in [ContentDB](https://github.com/foundpatterns/contentdb).

then, you could develop a lot further beyond.

**Comment by ithinuel** [comment]
@naturallymitchell maybe @saggit was simply looking for something to extract some data from a raw sql dump. Like a one-off log analysis tool. :D

**Comment by MarkMcCaskey** [comment]
I made a GameBoy ROM parser with nom5!
https://github.com/MarkMcCaskey/gameboy-rom-parser
https://crates.io/crates/gameboy-rom

It's extremely simple and doesn't do much, but the crate provides a useful abstraction over the metadata of GameBoy ROMs.

I'll add more optional validation functions to it and refactor my emulator's ROM code to use it soon.

**Comment by naturallymitchell** [comment]
> It's extremely simple and doesn't do much, but the crate provides a useful abstraction over the metadata of GameBoy ROMs.

@MarkMcCaskey  It could even make sense to refactor it then into a generalized library with config files (like, TOML and YAML, and now [SANE](https://gitlab.com/bloom42/sane-rs)).  Do you think that'd be too much more work?

**Comment by dwerner** [comment]
@Geal - I wanted to post my public suffix domain list parser that I wrote a few months back. I couldn't find a performant library that did what I needed, so I grabbed nom and went to work. https://github.com/dwerner/nom-psl

**Comment by MarkMcCaskey** [comment]
@naturallymitchell 

Do you mean specifying the layout of the bytes as data and creating a dynamic data structure from it?  That's an interesting idea, but I don't think it'd be too helpful for my use case -- as I see it, the primary value-add of the gameboy rom parser is the data layer that it exposes, which lets the user get things like the game's title as as string or the exact cartridge type and how much ROM and RAM it has as well-named, plain Rust values.

The parser may be implementable with serde deserialize on a `repr(C)` struct though, which is kind of the reverse of what you're saying, I think... I'm not familiar enough with how serde-derive handles errors though.

**Comment by o0Ignition0o** [comment]
Just got a 0.0.1 version of an NMEA-0183 parser using nom 5 https://github.com/YellowInnovation/nmea-0183 . I need to have a look at the docs and guidelines (the code is ugly for now) and refactor it :) I hope to submit a pull request adding a clean version of it to the parsers list soon ! :)

**Comment by armatusmiles** [comment]
This is SIP parser + validator. 
https://github.com/armatusmiles/sipcore/tree/master/crates/sipmsg

Torture test: https://github.com/armatusmiles/sipcore/commit/32040e57435a4d0cf2cb847520e8daafa4d8ad97 ( proof of work )

**Comment by Geal** [comment]
@armatusmiles thanks, i added it to the list in 2e58a2c

**Comment by bionicles** [comment]
Please please add OpenCypher to the list... a nice way to parse Graph DB queries could enable a wave of innovation in databases. There are zero legit serverless / autoscaling or decentralized graph databases (like you'd get with a CRDT/ORDT backend for an OpenCypher parser). GunJS is fairly close but JavaScript is not ideal for storage IMHO

**Comment by OtaK** [comment]
Wrote a [UBJSON](https://ubjson.org/) parser w/ nom
Pretty early version with just parsing, but it does the job.

https://github.com/OtaK/ubjson

https://crates.io/crates/ubjson

**Comment by NilsIrl** [comment]
There is a PDF parser here: https://github.com/J-F-Liu/lopdf (it requires using the `nom_parser` feature).

FWIW, with lopdf, the nom parser is much faster than the default parser

**Comment by atmnk** [comment]
I wrote a tool with its own [programming language](https://qalens.com/site/corr/jpl) using nom. [here](https://github.com/atmnk/corr) is source repo.

**Comment by erihsu** [comment]
gds2-parser released at https://crates.io/crates/gds2_io. BTW, my pull request tag is #1497

**Comment by manuschillerdev** [comment]
would it be feasible to write an ecmascript/typescript parser with nom as well? Or would the scope be too big for that?

**Comment by alexrsagen** [comment]
I have written 2 (public) parsers using nom which may be used as examples:
- BitTorrent/bencoding: https://github.com/alexrsagen/rustorrent/blob/4076d0ea689a950021164d2fdd412021519e7c68/src/bencode.rs
- MS-CFB: https://github.com/alexrsagen/rs-nomcfb (probably not a perfect implementation, but usable to parse Outlook .msg files and as a basic example for how to write parsers using nom)

**Comment by edg-l** [comment]
I made a bencode parser (the format used by .torrent files), https://github.com/edg-l/nom-bencode/

**Comment by LikeLakers2** [comment]
Hey there! I was wondering why the Rust parser on this list is `syn`? From what I can tell, `syn` does not use `nom` (although it might have in the past).

Since this is a list of examples of parsers built with `nom`, I don't see why we should be linking to `syn` here.

**Comment by OtaK** [comment]
@LikeLakers2 
> Hey there! I was wondering why the Rust parser on this list is `syn`? From what I can tell, `syn` does not use `nom` (although it might have in the past).
> 
> Since this is a list of examples of parsers built with `nom`, I don't see why we should be linking to `syn` here.

https://github.com/dtolnay/syn/issues/476

`syn` was using `nom` until v0.15, this issue was created 3 years before `syn` dropped its usage of `nom`. That's why it's still linked here. 

You're absolutely correct that it should be removed though.

**Comment by eatgrass** [comment]
`mdict-parser` is a parser library for .mdx dictionary format file
https://github.com/eatgrass/mdict-parser

**Comment by wjwei-handsome** [comment]
`crussmap` is a parser library and tool for .chain file [format](https://genome.ucsc.edu/goldenPath/help/chain.html)

https://github.com/wjwei-handsome/crussmap

**Comment by SergioRibera** [comment]
I have written a tool that allows you to render text templates, I leave you the following details

## Features
- Super fast
- Efficient
- Renders variables of all types
- Function system
- Easy implementation of custom functions
- Minimum possible dependencies
- Feature-based implementations, use only what you need
- Using the same variables renders multiple times

## Links
- Documentation [here](https://docs.rs/srtemplate/)
- [Wiki](https://github.com/SergioRibera/srtemplate/wiki)
- You can see a real example [here](https://sergioribera.github.io/srtemplate/), it is a real time template renderer

**Comment by ifsheldon** [comment]
I am a bit surprised that no one mentioned HTML!? I saw [nom_html_parser](https://gitlab.com/opentooladd/nom_html_parser), but it was long left unmaintained.

**Comment by YuniqueUnic** [comment]
[nomcurl](https://github.com/YuniqueUnic/nomcurl)

`nomcurl` is a Rust library and CLI dedicated to turning raw `curl …` strings into structured data. It now targets nom 8, exposes a stable parse_curl_command API, and keeps the CLI focused on fast inspection of individual parts of a request.
